### ISSUE TITLE
Epic: verified Qwen3 speculative inference to DeepSeek-V4 on gfx942

### ISSUE THREAD

**Issue by harsh-nod** [harsh-nod/ferric#1]
## Objective

Build Ferric as a proof-carrying, high-performance Rust inference engine with one production path:

```text
authenticated model bundle
  -> verified algorithm and execution plan
  -> proof-bound fe2o3 kernels
  -> generated Rust runner
  -> reviewed direct HSA command batches
```

The first product milestone is Qwen3-8B target inference with a pinned Qwen3-0.6B speculative draft model on one `gfx942`. The graduation target is DeepSeek-V4 speculative inference on eight `gfx942` devices.

Detailed contracts:

- [Roadmap](https://github.com/harsh-nod/ferric/blob/main/docs/ROADMAP.md)
- [Assurance policy](https://github.com/harsh-nod/ferric/blob/main/docs/ASSURANCE.md)
- [Architecture](https://github.com/harsh-nod/ferric/blob/main/docs/ARCHITECTURE.md)
- [Feature ledger](https://github.com/harsh-nod/ferric/blob/main/docs/FEATURES.md)
- [Performance protocol](https://github.com/harsh-nod/ferric/blob/main/docs/PERFORMANCE.md)

## Non-negotiable rules

- [ ] Keep one final execution path with no PyTorch, vendor-kernel, legacy-fe2o3, JIT, raw-launch, or plugin fallback.
- [ ] Put reusable compiler, kernel, proof, and HSA capabilities in fe2o3.
- [ ] Fail closed when a required feature, property status, identity, schedule, or target is unsupported.
- [ ] Report property-level evidence rather than one aggregate verified bit.
- [ ] Keep correctness and performance qualification independent and identity-bound.
- [ ] Pin exact vLLM and SGLang parity baselines before qualification.

## M0: permanent foundations

- [x] Create Ferric repository and Rust workspace.
- [x] Publish roadmap, assurance, architecture, feature, and performance contracts.
- [x] Implement executable greedy speculative-round semantics.
- [x] Implement exclusive generational paged-KV ownership.
- [x] Implement tentative append, commit, rollback, release, and stale-handle rejection.
- [ ] Implement generational request slots and deterministic scheduling.
- [ ] Implement completion epochs and retirement-before-reuse.
- [ ] Implement sealed prefix pages and copy-on-write.
- [ ] Prove M0 state machines in Verus and bind their property records.

### Build and model admission

- [ ] Define bounded canonical Ferric deployment bundles.
- [ ] Strictly admit Qwen3 config, tokenizer, vocabulary, and safetensors.
- [ ] Check target/draft tokenizer compatibility.
- [ ] Authenticate streamed prepacked weights.
- [ ] Generate an exact model/target runner and complete plan identity.

### Reusable fe2o3 kernel families

- [ ] BF16/FP32 GEMM and GEMV.
- [ ] RMSNorm, RoPE, residuals, and SwiGLU.
- [ ] FlashAttention prefill.
- [ ] GQA-aware paged decode attention and KV write.
- [ ] Logit projection, argmax, and compact completion records.
- [ ] Finite proof-bound schedules and property-preserving fusion.

### Reusable fe2o3 runtime work

- [ ] Reviewed long-lived HSA queue authority.
- [ ] Typed device and host-pinned allocation leases.
- [ ] Fixed-capacity asynchronous multi-packet command batches.
- [ ] One queue publication per generated inference step.
- [ ] Quiescence capability that retains every referenced resource.

### Ferric runtime

- [ ] Paged KV and continuous batching.
- [ ] Cancellation and quiescent retirement.
- [ ] Generated target-only Qwen3 runner.
- [ ] On-device draft, target verification, accept, commit, and rollback round.
- [ ] Greedy sampling and exact compact completion application.

### Proof and validation

- [ ] Model bundle well-formedness.
- [ ] Operator and generated graph refinement.
- [ ] Paged KV refinement to contiguous logical KV.
- [ ] Continuous batching refinement to independent requests.
- [ ] Greedy speculation refinement to target-only decoding.
- [ ] Request isolation, resource bounds, and completion-safe lifetimes.
- [ ] Kernel bounds, initialization, race, convergence, and numerical properties.
- [ ] Exact ABI, artifact, proof, target, and generated-plan identity closure.
- [ ] Publish the exact compiler/runtime/hardware TCB and last proved boundary.

### Performance qualification

- [ ] Differential logits/tokens and guarded hardware tests on `mi300x`.
- [ ] Cancellation, rollback, exhaustion, and fault-injection tests.
- [ ] Core kernels pass fe2o3 D10 gates.
- [ ] Speculation beats Ferric target-only execution on the eligible holdout.
- [ ] Compare with pinned tuned vLLM and SGLang at equal p99 SLO.
- [ ] Require the published faster claim to clear the 1.05 lower-confidence-bound gate.

## M2: serving and decoding breadth

- [ ] Counter-based domain-separated request RNG.
- [ ] Temperature, top-k, top-p, and stochastic speculation proof.
- [ ] Prefix caching, copy-on-write, and chunked prefill.
- [ ] Quantized KV, beam/parallel sampling, and structured output.
- [ ] LoRA and multi-LoRA batching.
- [ ] OpenAI-compatible serving boundary outside the proof core.

## M3: multi-device refinement

- [ ] Atomic multi-device execution epochs.
- [ ] Collective ordering, ownership, completion, and failure contracts.
- [ ] Tensor parallelism on two then eight `gfx942` devices.
- [ ] Data, pipeline, and expert parallelism.
- [ ] Distributed KV, prefix state, sampling, and speculative commit/rollback.
- [ ] Disaggregated prefill/decode state transfer.

## M4: DeepSeek-V4 graduation

- [ ] General top-k MoE routing, capacity, permutation, grouped GEMM, and combine.
- [ ] Expert placement, all-to-all, and expert-parallel refinement.
- [ ] FP8/FP4 numerical contracts and qualified fe2o3 kernels.
- [ ] DeepSeek-V4 CSA/HCA attention and mHC semantics.
- [ ] DeepSeek-V4-Flash target-only inference on eight `gfx942` devices.
- [ ] Distributed refinement for the admitted configuration.
- [ ] Apply the generic speculation protocol to a compatible pinned draft.
- [ ] Qualify DeepSeek-V4-Pro and longer-context configurations.

## Current evidence

Commit `f2e812f` passes formatting, clippy, and 13 Rust unit tests locally. The same source tree passes all 13 tests with Rust 1.75 on SSH host `mi300x`, which reports target `gfx942`.

No model, GPU kernel, HSA dispatch, verified-inference, machine-refinement, or performance claim exists yet.

**Comment by harsh-nod** [comment]
Foundation update: commits `f2e812f` through `4ed4ab4` are published on `main`. GitHub Actions is green on Rust 1.75. The same workspace passes formatting, clippy, and all 13 tests locally and over SSH on `mi300x` (`gfx942`). Implemented scope remains M0 executable semantics and KV metadata state only; no model/GPU/verified-inference/performance claim is being made.

**Comment by harsh-nod** [comment]
Second M0 swarm launched from main commit 5c6ad80 after hostile review rejected all three first-pass branches.

Accepted architecture:
- Rust 1.97.1 pinned to Verus 0.2026.08.02.b677dd5.
- Actual executable transition bodies live inside same-source verus! blocks.
- Ordinary Cargo and cargo-verus release builds were experimentally byte-identical.
- A deliberate executable-body mutation compiled but failed Verus.
- Strict jobs use fresh target directories and --no-cheating because verifier arguments are not in Cargo freshness fingerprints.
- Shipped artifacts contain no Verus/vstd/Z3 runtime dependency.

Active worktrees:
- agent/m0-verified-lifecycle: unified Ready/InFlight/Retiring lifecycle, rotating scheduler, exact-next completion, quiescent reclaim.
- agent/m0-verified-kv: fixed arrays, page-aligned prefix sharing, initialized-read bounds, epoch-gated reclaim.
- agent/m0-proof-gate: direct identity/speculation proofs plus authenticated cargo-verus release receipt and CI.

Rejected commits remain unmerged and their issue comments record the blocking findings.

**Comment by harsh-nod** [comment]
M0 is complete and receipt-bound at `33323ba8f29ec431e8177eb3a636ddd64d1bebe9`.

Final gate:
- 527 Verus queries, 0 errors
- 117 directly verified admitted bodies; 123/123 executable bodies accounted for
- 37/37 same-source negative mutations rejected
- hostile policies 18/18 and 17/17
- 10 Proved, 1 Contracted, 2 Checked, and 2 Unsupported property records accepted by fe2o3 `ContractSetV1::validate_closed`
- strict Clippy, 39/39 debug tests, and 39/39 release tests
- exact-source `mi300x` CPU validation passed; no GPU workload was invoked
- receipt SHA-256: `6989d398aec00401f8d2196aa5b2802f8a66c720fc015269774766e5e913d4ac`

Closed M0 children: #2, #3, #4. This epic remains open for M1 onward. The published non-claims remain in force: no model runner, GPU kernels, HSA runtime binding, serving path, machine refinement, or performance result yet.

**Comment by harsh-nod** [comment]
Post-push confirmation: [standard CI](https://github.com/harsh-nod/ferric/actions/runs/32415873481) and the [authenticated Verus release workflow](https://github.com/harsh-nod/ferric/actions/runs/32415873506) both passed on exact commit `33323ba`.

The retained CI artifact `ferric-verus-qualification` contains its verified receipt, proof/runtime transcripts, property contracts/evidence, artifacts, and all mutation/policy transcripts. Downloaded CI receipt SHA-256: `f70dd2b0fc19da5818128d84cb91966590538dbf66d8d7c2d296126a2a83e058`; its companion checksum verifies successfully.
